netfox/Core/NFXProtocol.swift

```diff
@@ -54,99 +55,71 @@ open class NFXProtocol: URLProtocol
+        responseData?.append(data)
+        
         client?.urlProtocol(self, didLoad: data)
     }
     
     public func urlSession(_ session: URLSession, dataTask: URLSessionDataTask, didReceive response: URLResponse, completionHandler: @escaping (URLSession.ResponseDisposition) -> Void) {
-        let policy = URLCache.StoragePolicy(rawValue: request.cachePolicy.rawValue) ?? .notAllowed
-        client?.urlProtocol(self, didReceive: response, cacheStoragePolicy: policy)
+        self.response = response
+        self.responseData = NSMutableData()
+        
+        client?.urlProtocol(self, didReceive: response, cacheStoragePolicy: NFX.swiftSharedInstance.cacheStoragePolicy)
         completionHandler(.allow)
     }
     
     public func urlSession(_ session: URLSession, task: URLSessionTask, didCompleteWithError error: Error?) {
-        if let error = error {
-            client?.urlProtocol(self, didFailWithError: error)
-        } else {
-            client?.urlProtocolDidFinishLoading(self)
+        defer {
+            if let error = error {
+                client?.urlProtocol(self, didFailWithError: error)
+            } else {
+                client?.urlProtocolDidFinishLoading(self)
+            }
         }
-    }
-    
-    public func urlSession(_ session: URLSession, task: URLSessionTask, willPerformHTTPRedirection response: HTTPURLResponse, newRequest request: URLRequest, completionHandler: @escaping (URLRequest?) -> Void) {
-        if let mutableRequest = (request as NSURLRequest).mutableCopy() as? NSMutableURLRequest {
-            URLProtocol.removeProperty(forKey: "NFXInternal", in: mutableRequest)
-            client?.urlProtocol(self, wasRedirectedTo: request, redirectResponse: response)
-            completionHandler(request)
+        
+        guard let request = task.originalRequest else {
+            return
+        }
+        
+        model.saveRequestBody(request)
+        model.logRequest(request)
+        
+        if error != nil {
+            model.saveErrorResponse()
+        } else if let response = response {
+            let data = (responseData ?? NSMutableData()) as Data
+            model.saveResponse(response, data: data)
         }
+        
+        NFXHTTPModelManager.sharedInstance.add(model)
+        NotificationCenter.default.post(name: .NFXReloadData, object: nil)
     }
     
-    public func urlSession(_ session: URLSession, didBecomeInvalidWithError error: Error?) {
-        guard let error = error else { return }
-        client?.urlProtocol(self, didFailWithError: error)
+    public func urlSession(_ session: URLSession, task: URLSessionTask, willPerformHTTPRedirection response: HTTPURLResponse, newRequest request: URLRequest, completionHandler: @escaping (URLRequest?) -> Void) {
+        client?.urlProtocol(self, wasRedirectedTo: request, redirectResponse: response)
+        completionHandler(request)
```

Comment:
Question about this, as we start loading, we should not continue removing before call our completion handler?
URLProtocol.removeProperty(forKey: NFXProtocol.nfxInternalKey, in: mutableRequest)

Reply:
I was wondering if it was needed to remove it, I guess it would be better for when people reuse the request from the `URLTask`, I'll put it back.

Reply:
Looking at it again, I think it is done, making sure the property does not get forwarded to the final request of the `URLDataTask` being send to the original caller of the request. I'll put it back.